clients/src/main/java/org/apache/kafka/clients/consumer/internals/Heartbeat.java

```diff
@@ -16,26 +16,41 @@
  * A helper class for managing the heartbeat to the coordinator
  */
 public final class Heartbeat {
-    private final long timeout;
-    private final long interval;
+    private final long sessionTimeout;
+    private final long heartbeatInterval;
+    private final long maxPollInterval;
+    private final long retryBackoffMs;
 
-    private long lastHeartbeatSend;
+    private volatile long lastHeartbeatSend; // volatile since it is read by metrics
     private long lastHeartbeatReceive;
     private long lastSessionReset;
+    private long lastPoll;
+    private boolean heartbeatFailed;
 
-    public Heartbeat(long timeout,
-                     long interval,
-                     long now) {
-        if (interval >= timeout)
+    public Heartbeat(long sessionTimeout,
+                     long heartbeatInterval,
+                     long maxPollInterval,
+                     long retryBackoffMs) {
+        if (heartbeatInterval >= sessionTimeout)
             throw new IllegalArgumentException("Heartbeat must be set lower than the session timeout");
 
-        this.timeout = timeout;
-        this.interval = interval;
-        this.lastSessionReset = now;
+        this.sessionTimeout = sessionTimeout;
+        this.heartbeatInterval = heartbeatInterval;
+        this.maxPollInterval = maxPollInterval;
+        this.retryBackoffMs = retryBackoffMs;
+    }
+
+    public void poll(long now) {
+        this.lastPoll = now;
     }
 
     public void sentHeartbeat(long now) {
         this.lastHeartbeatSend = now;
+        this.heartbeatFailed = false;
+    }
+
+    public void failHeartbeat() {
```

Comment:
@guozhangwang I don't think it's a problem to backoff an attempted heartbeat since the heartbeat interval shouldn't be set close to the session timeout in practice, but it's reasonable to ask whether it's necessary. 

Most of the usual coordinator error codes are possible from the heartbeat response. Usually we require some other action before we can continue sending heartbeats. For example, we may need to rediscover the coordinator if we receive NOT_COORDINATOR_FOR_GROUP, or we may need to rejoin the group if we receive ILLEGAL_GENERATION. We may also get a socket disconnect or a request timeout, and this is when the backoff is useful. What often happens, when the coordinator is shutdown, is that we'll immediately get the disconnect on the client, but it will take some time for the cluster to determine who the next coordinator will be and propagate that information to all brokers. So we can end up doing the following in a fairly tight loop:

1. Find the current coordinator -> returns old coordinator
2. Attempt to send heartbeat to old coordinator -> disconnect
3. Go back to 1.

We eventually break from the loop when the coordinator change has been propagated and step 1. returns its id, but it can take several long seconds for this to happen. The backoff is helpful in this case because it controls the rate of the loop. So I think the question is whether we want to try to distinguish this case so that we only backoff when it's encountered, or we just backoff for any failure. I chose the latter for simplicity and because it didn't seem to have any bad effect. I'd be interested to hear your thoughts.